keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+All Kubernetes APIs support [field
+selection](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/)
+for metadata.name, metadata.namespace.
+
+Additionally, built-in APIs support field selection on specific fields such as
+status.phase for Pods, but CustomResourceDefinitions lack equivalent
+functionality.
+
+Without this enhancement, an extension author that wants to be be able to filter
+resources might choose to:
+
+  - Organize data into a label even though it should be a spec or status field.
+  - Double writing data both to a label and a spec or status field.
+
+Both of these should be avoided. We should enable extension authors structure
+their APIs according to best practices, not based on data access patterns.
+
+### Goals
+
+- Add an allow list of selectable fields to CustomResourceDefinitions.
+- Performance and resource consumption characteristics are roughtly equivalent
+  to label selectors.
+
+### Non-Goals
+
+- Arbitrary field selection.
+   - This proposal is to support field selection for a small allow list of
+     fields. Expanding support to include all fields is complicated by:
+     - Fields contained in, or nested under, lists or maps.
+     - The costs to apiserver of evaluating filters on objects or maintaining
+       indices.
+     - Lack of direct support for filtering in etcd.
+- More sophisticated field selection logic. Possible approaches include: CEL
+  based field selection, enhancements to field selection such as set based
+  operators like provided for labels ("in", "not in"), OR clauses, pattern
```

Comment:
in and not in are supported for label selectors today.  While I understand not expanding to encompass features like pattern matching (not in label selectors), why continue excluding in and not in?

Reply:
It's a gap in how `--field-selector` is implemented today.   I'm not opposed to "in" and "not in" support, it's just an improvement I wasn't intending to tackle with this KEP.  It would be fiddly to add.. similar to how we add functions to CEL.

We will cap out on expressiveness pretty quick field selectors, but maybe complementing the existing AND functionality with "in" and "not in" would be a good finishing line for field selectors?  Should we try for it in beta?

Reply:
For me personally those are separate things.
I'm definitely +1 for supporting in/not-in, but this is independent from supportin field-selectors for CRDs that this KEP is about.

So I would suggest separating these two and tracking in/not-in as a separate KEP.